include/revng/Support/GraphAlgorithms.h

```diff
@@ -196,3 +201,215 @@ auto exitless_scc_range(NodeTy Entry) {
 
   return make_filter_range(Range, Filter);
 }
+
+namespace detail {
+
+template<typename T>
+T stripInverseHelper(llvm::Inverse<T> &&);
+
+template<typename T>
+T stripInverseHelper(T &&);
+
+template<typename T>
+using StripInverse = decltype(stripInverseHelper(std::declval<T>()));
+
+// clang-format off
+template<typename T>
+concept SupportedGraph = std::is_pointer_v<T> ||
+                         (StrictSpecializationOf<T, llvm::Inverse> &&
+                          std::is_pointer_v<StripInverse<T>>);
+// clang-format on
+
+template<typename GraphType>
+using GT = llvm::GraphTraits<GraphType>;
+
+} // namespace detail
```

Comment:
I don't think that's possible, since this concept (and everything but `GT` is a part of it) is here  to specialize the `GraphNodeOrRootHelper` class itself. Hence it needs to be explicitly defined before the template definition.

On the other hand, I don't feel like these requirements are wide enough to promote them into one of the higher level headers (like `ADT/Concepts.h` or something).

Reply:
Uh right, I'm so sad we can't have `concept`s inside a `class` or `struct`.

But I wouldn't promote them to any higher level headers yet. I really want to be extra sure that we need all this machinery, because I still feel there should be a way to avoid this `GraphNodeOrRootHelper` alltogether.

Reply:
I'm closing this conversation for now. Leave stuff in `detail`. If we deem it worthy to be promoted in a higher level header we'll move it later, after we've resolved the other discussion on `GraphNodeOrRootHelper`.